### PR TITLE
Support slices and converting raw data into data model

### PR DESCRIPTION
**Pull request by ekoval**
Required for: https://github.com/yola/importservice/pull/189

Usage example: 
```
class PageOfAssets(Page):
    _data_model_class = Asset

    @property
    def items(self):
        raw_items = super().items
        return [self._data_model_class(**d) for d in raw_items]
...
...

    @classmethod
    def list(cls, **attributes):

        return PaginatedResults(
            cls._get_ws_imports_client().list_assets, kwargs=attributes,
            page_size=20, page_class=PageOfAssets
        )
```

### CODE REVIEW

**Comment by jamesodo** [comment]
Eugene and I discussed this in slack. He said he tried a differently implementation where `PaginatedResults` accepted a parameter for a `Page` subclass instead of `data_model_class`.
The `Page` subclass could then override the `items` property.
I think I like that way better.

File: demands/pagination.py [hunk 1]
```diff
@@ -120,29 +152,13 @@ def _page_ids(self):
...
-        next_page_is_null = (
-            next_key in self._data and
-            self._data[next_key] is None
-        )
-        return self.size < self._options[PAGE_SIZE] or next_page_is_null
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            page_num = int(key.start / self.options['page_size']) + 1
+            return self._get_page(page_num).items
+        return self.list[key]
```
Comment (RayeN) [inline demands/pagination.py]: What is `self.list`?

File: demands/pagination.py [hunk 2]
```diff
@@ -120,29 +152,13 @@ def _page_ids(self):
             return count(start, self.options[PAGE_SIZE])
         raise ValueError('Unknown pagination_type')
 
+    def count(self):
+        if self._count is None:
+            self._count = self._get_page(1)._data['count']
```
Comment (RayeN) [inline demands/pagination.py]: It won't work for all cases.

File: demands/pagination.py [hunk 3]
```diff
@@ -120,29 +152,13 @@ def _page_ids(self):
...
-        next_page_is_null = (
-            next_key in self._data and
-            self._data[next_key] is None
-        )
-        return self.size < self._options[PAGE_SIZE] or next_page_is_null
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            page_num = int(key.start / self.options['page_size']) + 1
+            return self._get_page(page_num).items
+        return list(self)[key]
```
Comment (RayeN) [inline demands/pagination.py]: I think this can be optimized.

**Comment by RayeN** [comment]
Hmm... I don't think it's a responsibility of this library to be aware of data wrapping.  I think, in your example, you mixed service and model levels.  Usually, we have service method which returns us PaginatedResults (iterable) and model which iterates those results and wraps them into the appropriate object.

**Comment by ekoval** [comment]
> Hmm... I don't think it's a responsibility of this library to be aware of data wrapping. I think, in your example, you mixed service and model layers. Usually, we have service method which returns us PaginatedResults (iterable) and model which iterates those results and wraps them into the appropriate object.

It is not responsible for data wrapping. Demands provides an ability to pass custom `Page` class, which in turn cares about data.

File: demands/pagination.py (hunk 2 above)
Comment (ekoval) [reply demands/pagination.py]: It would be good to say why :)

**Comment by RayeN** [comment]
> It is not responsible for data wrapping.

You're right.  But we're making changes to demands to allow to do that in custom `Page`, which, IMO, will mix service and model layers.

File: demands/pagination.py (hunk 2 above)
Comment (RayeN) [reply demands/pagination.py]: for example when there's no 'count' key

Comment (ekoval) [reply demands/pagination.py]: This is not a prod code, we're discussing an idea. If it's ok, I'll polish it.

**Comment by RayeN** [comment]
> Eugene and I discussed this in slack. He said he tried a differently implementation where PaginatedResults accepted a parameter for a Page subclass instead of data_model_class.
The Page subclass could then override the items property.
I think I like that way better.

And I liked the initial version with a generic wrapper which (if passed) Page will use to wrap raw data. Because clients of demands won't need to know anything about the internal implementation of `PaginatedResults` nor they need to mix `Page` and `Model` layers while overriding `Page` implementation.

**Comment by ekoval** [comment]
To reach a compromise, we can give that `_data_model_class` a more generic name. What about `wrapper`, @jamesodo ?

**Comment by jamesodo** [comment]
I think this may be a case where we're trying to do too much with the models.
The `WSImportsClient().list_*` service client methods already give all the info we need to do what this is trying to accomplish.

**Comment by ekoval** [comment]
You mean just proxy `list()` endpoints transparently and let WSi paginate data? It's not that easy I guess. At least prev/next links will be broken.

We also can't just return raw data to `IS.get_queryset` - it requires queryset which consists of models.

**Comment by RayeN** [comment]
We should probably make a decision whether we need these changes or not in some other repo (just a reminder - it's a public repo).  Despite our needs, I like the ideas of wrapping raw data, slices and `__len__()`.

**Comment by blaix** [comment]
I think we can close this?

**Pull request closed without merging**